File: commons/version/src/test/java/com/cedarsoft/version/VersionTest.java
/**
 * Copyright (C) cedarsoft GmbH.
 *
 * Licensed under the GNU General Public License version 3 (the "License")
 * with Classpath Exception; you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *         http://www.cedarsoft.org/gpl3ce
 *         (GPL 3 with Classpath Exception)
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 only, as
 * published by the Free Software Foundation. cedarsoft GmbH designates this
 * particular file as subject to the "Classpath" exception as provided
 * by cedarsoft GmbH in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 3 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 3 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact cedarsoft GmbH, 72810 Gomaringen, Germany,
 * or visit www.cedarsoft.com if you need additional information or
 * have any questions.
 */

package com.cedarsoft.version;

import static org.assertj.core.api.Assertions.*;

import org.junit.jupiter.api.*;

/**
 *
 */
public class VersionTest {
  @Test
  public void testSmapper() throws Exception {
    assertThat(Version.valueOf(1, 0, 0).smallerThan(Version.valueOf(2, 0, 0))).isTrue();
    assertThat(Version.valueOf(3, 1, 0, "SNAPSHOT").smallerThan(Version.valueOf(2, 0, 0))).isFalse();
    assertThat(Version.valueOf(2, 0, 0).smallerThan(Version.valueOf(3, 1, 0, "SNAPSHOT"))).isTrue();
    assertThat(Version.valueOf(3, 0, 0).smallerThan(Version.valueOf(3, 1, 0, "SNAPSHOT"))).isTrue();
  }

  @Test
  public void parseLongNumber() throws Exception {
    Version version = Version.parse("0.0.110");
    assertThat(version.getMajor()).isEqualTo(0);
    assertThat(version.getMinor()).isEqualTo(0);
    assertThat(version.getBuild()).isEqualTo(110);
  }

  @Test
  public void testMatch() {
    Version.verifyMatch( Version.valueOf( 1, 2, 3 ), Version.valueOf( 1, 2, 3 ) );

    try {
      Version.verifyMatch(Version.valueOf(1, 2, 3), Version.valueOf(1, 2, 4));
      Assertions.fail("Where is the Exception");
    }
    catch (VersionMismatchException ignore) {
    }
  }

  @Test
  public void testToInt() {
    Assertions.assertEquals(10203, new Version(1, 2, 3).toInt());
    Assertions.assertEquals(123456, new Version(12, 34, 56).toInt());
  }

  @Test
  public void testEquals() {
    Assertions.assertEquals(new Version(1, 2, 3), new Version(1, 2, 3));
    Assertions.assertEquals(new Version(1, 2, 3, "asdf"), new Version(1, 2, 3, "asdf"));
  }

  @Test
  public void testToString() {
    Assertions.assertEquals("1.2.3", new Version(1, 2, 3).toString());
    Assertions.assertEquals("1.2.3-asdf", new Version(1, 2, 3, "asdf").toString());
  }

  @Test
  public void testParse() {
    Assertions.assertEquals(new Version(1, 2, 3), Version.parse("1.2.3"));
    Assertions.assertEquals(new Version(1, 2, 3, "build76"), Version.parse("1.2.3-build76"));
    Assertions.assertEquals(new Version(1, 2, 3, "build76"), Version.parse(new Version(1, 2, 3, "build76").toString()));
  }

  @Test
  public void testCompareGreater() {
    Assertions.assertTrue(new Version(1, 2, 3).sameOrGreaterThan(new Version(1, 2, 3)));
    Assertions.assertTrue(new Version(1, 2, 3).sameOrGreaterThan(new Version(1, 2, 2)));
    Assertions.assertTrue(new Version(1, 2, 3).sameOrGreaterThan(new Version(0, 2, 2)));

    Assertions.assertFalse(new Version(1, 2, 3).sameOrGreaterThan(new Version(1, 2, 4)));
  }

  @Test
  public void testCompareSmaller() {
    Assertions.assertTrue(new Version(1, 2, 3).sameOrSmallerThan(new Version(1, 2, 3)));
    Assertions.assertTrue(new Version(1, 2, 3).sameOrSmallerThan(new Version(1, 2, 4)));
    Assertions.assertTrue(new Version(1, 2, 3).sameOrSmallerThan(new Version(2, 2, 2)));

    Assertions.assertFalse(new Version(1, 2, 3).sameOrSmallerThan(new Version(0, 2, 4)));
  }

  @Test
  public void withMethods() throws Exception {
    Version version = Version.valueOf(1, 2, 3);
    assertThat(version.withSuffix("daSuffix").getSuffix()).isEqualTo("daSuffix");
  }
}

File: commons/version/src/jvmTest/kotlin/com/cedarsoft/version/VersionTest.kt
package com.cedarsoft.version

import com.cedarsoft.version.Version.Companion.parse
import com.cedarsoft.version.Version.Companion.valueOf
import com.cedarsoft.version.Version.Companion.verifyMatch
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assertions.fail
import org.junit.jupiter.api.Test

/**
 *
 */
class VersionTest {
  @Test
  @Throws(Exception::class)
  fun testSmapper() {
    Assertions.assertThat(valueOf(1, 0, 0).smallerThan(valueOf(2, 0, 0))).isTrue
    Assertions.assertThat(valueOf(3, 1, 0, "SNAPSHOT").smallerThan(valueOf(2, 0, 0))).isFalse
    Assertions.assertThat(valueOf(2, 0, 0).smallerThan(valueOf(3, 1, 0, "SNAPSHOT"))).isTrue
    Assertions.assertThat(valueOf(3, 0, 0).smallerThan(valueOf(3, 1, 0, "SNAPSHOT"))).isTrue
  }

  @Test
  @Throws(Exception::class)
  fun parseLongNumber() {
    val version = parse("0.0.110")
    Assertions.assertThat(version.major).isEqualTo(0)
    Assertions.assertThat(version.minor).isEqualTo(0)
    Assertions.assertThat(version.build).isEqualTo(110)
  }

  @Test
  fun testMatch() {
    verifyMatch(valueOf(1, 2, 3), valueOf(1, 2, 3))
    try {
      verifyMatch(valueOf(1, 2, 3), valueOf(1, 2, 4))
      fail<Any>("Where is the Exception")
    } catch (ignore: VersionMismatchException) {
    }
  }

  @Test
  fun testToInt() {
    assertEquals(10203, Version(1, 2, 3).toInt())
    assertEquals(123456, Version(12, 34, 56).toInt())
  }

  @Test
  fun testEquals() {
    assertEquals(Version(1, 2, 3), Version(1, 2, 3))
    assertEquals(Version(1, 2, 3, "asdf"), Version(1, 2, 3, "asdf"))
  }

  @Test
  fun testToString() {
    assertEquals("1.2.3", Version(1, 2, 3).toString())
    assertEquals("1.2.3-asdf", Version(1, 2, 3, "asdf").toString())
  }

  @Test
  fun testParse() {
    assertEquals(Version(1, 2, 3), parse("1.2.3"))
    assertEquals(Version(1, 2, 3, "build76"), parse("1.2.3-build76"))
    assertEquals(Version(1, 2, 3, "build76"), parse(Version(1, 2, 3, "build76").toString()))
  }

  @Test
  fun testCompareGreater() {
    assertTrue(Version(1, 2, 3).sameOrGreaterThan(Version(1, 2, 3)))
    assertTrue(Version(1, 2, 3).sameOrGreaterThan(Version(1, 2, 2)))
    assertTrue(Version(1, 2, 3).sameOrGreaterThan(Version(0, 2, 2)))
    assertFalse(Version(1, 2, 3).sameOrGreaterThan(Version(1, 2, 4)))
  }

  @Test
  fun testCompareSmaller() {
    assertTrue(Version(1, 2, 3).sameOrSmallerThan(Version(1, 2, 3)))
    assertTrue(Version(1, 2, 3).sameOrSmallerThan(Version(1, 2, 4)))
    assertTrue(Version(1, 2, 3).sameOrSmallerThan(Version(2, 2, 2)))
    assertFalse(Version(1, 2, 3).sameOrSmallerThan(Version(0, 2, 4)))
  }

  @Test
  @Throws(Exception::class)
  fun withMethods() {
    val version = valueOf(1, 2, 3)
    Assertions.assertThat(version.withSuffix("daSuffix").suffix).isEqualTo("daSuffix")
  }

}
